src/darwin/Framework/CHIP/CHIPDevice.mm

```diff
@@ -173,6 +202,898 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
+                         clientQueue:(dispatch_queue_t)clientQueue
+                          completion:(CHIPDeviceResponseHandler)completion
+{
+    new NSObjectAttributeCallbackBridge(
+        clientQueue, completion, ^(chip::Callback::Cancelable * success, chip::Callback::Cancelable * failure) {
+            auto successFn = chip::Callback::Callback<NSObjectAttributeCallback>::FromCancelable(success);
+            auto failureFn = chip::Callback::Callback<CHIPErrorCallback>::FromCancelable(failure);
+            auto context = successFn->mContext;
+            auto successCb = successFn->mCall;
+            auto failureCb = failureFn->mCall;
+            auto resultArray = [[NSMutableArray alloc] init];
+            auto resultSuccess = [[NSMutableArray alloc] init];
+            auto resultFailure = [[NSMutableArray alloc] init];
+            auto onSuccessCb = [resultArray, resultSuccess](const app::ConcreteAttributePath & commandPath) {
+                [resultArray
+                    addObject:[[NSDictionary alloc] initWithObjectsAndKeys:[NSNumber numberWithUnsignedInt:commandPath.mEndpointId],
+                                                    kCHIPEndpointIdKey, [NSNumber numberWithUnsignedInt:commandPath.mClusterId],
+                                                    kCHIPClusterIdKey, [NSNumber numberWithUnsignedInt:commandPath.mAttributeId],
+                                                    kCHIPAttributeIdKey, [NSNumber numberWithUnsignedInt:0], kCHIPStatusKey, nil]];
+                if ([resultSuccess count] == 0) {
+                    [resultSuccess addObject:[NSNumber numberWithBool:YES]];
+                }
+            };
+
+            auto onFailureCb = [resultArray, resultFailure](const app::ConcreteAttributePath * commandPath, CHIP_ERROR aError) {
+                if (commandPath) {
+                    [resultArray addObject:[[NSDictionary alloc]
+                                               initWithObjectsAndKeys:[NSNumber numberWithUnsignedInt:commandPath->mEndpointId],
+                                               kCHIPEndpointIdKey, [NSNumber numberWithUnsignedInt:commandPath->mClusterId],
+                                               kCHIPClusterIdKey, [NSNumber numberWithUnsignedInt:commandPath->mAttributeId],
+                                               kCHIPAttributeIdKey, [NSNumber numberWithUnsignedInteger:aError.AsInteger()],
+                                               kCHIPStatusKey, nil]];
+                } else {
+                    if ([resultFailure count] == 0) {
+                        [resultFailure addObject:[NSNumber numberWithUnsignedInteger:aError.AsInteger()]];
+                    }
+                }
+            };
+
+            auto onDoneCb = [context, successCb, failureCb, resultArray, resultSuccess, resultFailure](
+                                app::WriteClient * pWriteClient) {
+                if ([resultFailure count] > 0 || [resultSuccess count] == 0) {
```

Comment:
I wonder whether we can reduce the code duplication with the read case here somehow....